apps/challenges/aws_utils.py

```diff
@@ -899,6 +899,339 @@ def create_eks_nodegroup(challenge, cluster_name):
     create_service_by_challenge_pk(client, challenge_obj, client_token)
 
 
+@app.task
+def delete_challenge_evaluation_cluster_and_roles(queryset):
+    """
+    The function called by the admin action method to delete all the challenge evaluation clusters used by the
+    selected challenges. Calls the delete_eks_cluster_and_roles method.
+
+    Parameters:
+    queryset (<class 'django.db.models.query.QuerySet'>): The queryset of selected challenges in the django admin page.
+
+    Returns:
+    dict: keys-> 'count': the number of workers successfully stopped.
+                 'failures': a dict of all the failures with their error messages and the challenge pk
+    """
+    from .models import ChallengeEvaluationCluster
+
+    if settings.DEBUG:
+        failures = []
+        for challenge in queryset:
+            failures.append(
+                {
+                    "message": "Challenge evaluation clusters cannot be deleted in development environment",
+                    "challenge_pk": challenge.pk,
+                }
+            )
+        return {"count": 0, "failures": failures}
+
+    count = 0
+    failures = []
+    for challenge in queryset:
+        try:
+            ChallengeEvaluationCluster.objects.get(challenge=challenge)
+        except ChallengeEvaluationCluster.DoesNotExist:
+            response = "Please select challenges with active evaluation clusters only."
+            failures.append(
+                {"message": response, "challenge_pk": challenge.pk}
+            )
+            continue
+        response = delete_eks_cluster_and_roles(challenge=challenge)
+        if not response or response["ResponseMetadata"]["HTTPStatusCode"] != HTTPStatus.OK:
+            failures.append(
+                {
+                    "message": response["Error"],
+                    "challenge_pk": challenge.pk,
+                }
+            )
+            continue
+        count += 1
+    return {"count": count, "failures": failures}
+
+
+@app.task
+def delete_eks_cluster_and_roles(challenge):
+    """
+    Deletes EKS and NodeGroup ARN roles
+
+    Arguments:
+        instance {<class 'django.db.models.query.QuerySet'>} -- instance of the model calling the post hook
+    """
+    from .models import ChallengeEvaluationCluster
+    from .utils import get_aws_credentials_for_challenge
+
+    for obj in serializers.deserialize("json", challenge):
+        challenge_obj = obj.object
+    challenge_aws_keys = get_aws_credentials_for_challenge(challenge_obj.pk)
+    client = get_boto3_client("iam", challenge_aws_keys)
+    environment_suffix = "{}-{}".format(challenge_obj.pk, settings.ENVIRONMENT)
+    eks_role_name = "evalai-code-upload-eks-role-{}".format(environment_suffix)
+    node_group_role_name = "evalai-code-upload-nodegroup-role-{}".format(
+        environment_suffix
+    )
+    try:
+        # Delete all EKS cluster subnets
+        delete_eks_cluster_subnets.delay(challenge)
```

Comment:
@matthew-so subnets need to be deleted at the very end. It won't work otherwise. The flow for deletion will be:

1. Delete EKS nodegroup
2. Delete EKS cluster
3. Delete EFS associated to challenge
4. Delete policies and IAM roles for the challenge
5. Delete subnets
6. Delete VPC

Can you restructure the calls based on this flow?

Reply:
Will do